### PR TITLE
[invariant] Assertion functions

### PR DESCRIPTION
**Pull request by josh**
Make use of TypeScript 3.7 assertion function feature to implement `invariant`.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions>

### CODE REVIEW

**Comment by bennett000** [comment]
Looks good to me.

That said I'm not a reviewer (on the right sidebar) and I cannot seem to approve as @typescript-bot suggests

File: types/invariant/invariant-tests.ts [hunk 1]
```diff
@@ -21,5 +21,5 @@ invariant('', 'Some other error');
 // handles extra variables
 invariant(true, 'Error, error, read all about it', 37, {}, 'hello');
 
-// $ExpectType {}
+// $ExpectType void | {}
```
Comment (dtinth) [inline types/invariant/invariant-tests.ts]: I believe this should stay as `{}` to facilitate this pattern:

```ts
const thing = getThing() || invariant(false, 'Expected getThing to return something')
```

as well as

```ts
const thing = getThing() ?? invariant(false, 'Expected getThing to return something')
```

File: types/invariant/index.d.ts [hunk 2]
```diff
@@ -11,8 +13,7 @@ export as namespace invariant;
 
 declare namespace invariant {
   interface InvariantStatic {
-    (testValue:false, format?:string, ...extra:any[]):never;
```
Comment (dtinth) [inline types/invariant/index.d.ts]: IMO we should keep this overload to not break existing test (see comment below).

File: types/invariant/index.d.ts [hunk 3]
```diff
@@ -11,8 +13,7 @@ export as namespace invariant;
...
-    (testValue:any, format?:string, ...extra:any[]):void;
+    (testValue:any, format?:string, ...extra:any[]):asserts testValue;
```
Comment (dtinth) [inline types/invariant/index.d.ts]: Would be great if some test can be added for this.

**Review by dtinth — CHANGES_REQUESTED** [review]
Thanks for the PR!

I'm excited to use assertion functions in TypeScript as well. However I think we should retain support for `invariant (false)` pattern.

**Comment by josh** [comment]
> I believe this should stay as `{}` to facilitate this pattern:

Is it possible for an `asserts` function to have a return value?

> However I think we should retain support for `invariant (false)` pattern.

👍 that makes sense.

**Review by dtinth — COMMENTED** [review]
@josh Thanks! Mind adding a corresponding test for this change? Maybe a test could look like this:

```ts
declare const x: any
invariant(typeof x === 'string')
const s = x
// $ExpectType string
x
```

**Comment by josh** [comment]
> Mind adding a corresponding test for this change? Maybe a test could look like this:

Tried that, but got:

```
expect  TypeScript@next expected type to be:
  string
got:
  any
```

**Pull request closed without merging**